src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Renamer.cs

```diff
@@ -0,0 +1,177 @@
+
+                    string oldName = Path.GetFileNameWithoutExtension(_oldDocument.FilePath);
+                    string newName = Path.GetFileNameWithoutExtension(newDocument.FilePath);
+
+                    var declaration = root.DescendantNodes().Where(n => HasMatchingSyntaxNode(n, oldName)).FirstOrDefault();
+                    if (declaration == null)
+                        return;
+
+                    var semanticModel = newDocument.GetSemanticModelAsync().Result;
+                    if (semanticModel == null)
+                        return;
+
+                    var symbol = semanticModel.GetDeclaredSymbol(declaration);
+                    if (symbol == null)
+                        return;
+
+                    // We're about to do a symbolic rename.If the user has asked us to prompt, We need to open a dialog and ask them  
+                    //  Otherwise go ahead and do the rename.
+                    EnvDTE.DTE dte = _serviceProvider.GetService<EnvDTE.DTE, EnvDTE.DTE>();
+                    var props = dte.Properties["Environment", "ProjectsAndSolution"];
+                    if (props != null)
+                    {
+                        if ((bool)props.Item("PromptForRenameSymbol").Value)
+                        {
+                            string promptMessage = string.Format(Resources.RenameSymbolPrompt, oldName);
+                            var result = VsShellUtilities.ShowMessageBox(_serviceProvider, promptMessage, null, OLEMSGICON.OLEMSGICON_QUERY,
+                                                            OLEMSGBUTTON.OLEMSGBUTTON_YESNO, OLEMSGDEFBUTTON.OLEMSGDEFBUTTON_FIRST);
+                            if (result == (int)VSConstants.MessageBoxResult.IDNO)
+                            {
+                                return;
+                            }
+                        }
+                    }
+
+                    // Now do the rename
+                    var optionSet = newDocument.Project.Solution.Workspace.Options;
+                    var renamedSolution = RoslyRenamer.Renamer.RenameSymbolAsync(newDocument.Project.Solution, symbol, newName, optionSet).Result;
+
+                    await _threadingService.SwitchToUIThread();
+
+                    if (!newDocument.Project.Solution.Workspace.TryApplyChanges(renamedSolution))
+                    {
+                        string promptMessage = string.Format(Resources.RenameSymbolFailed, oldName);
+                        var result = VsShellUtilities.ShowMessageBox(_serviceProvider, promptMessage, null, OLEMSGICON.OLEMSGICON_WARNING,
+                                                OLEMSGBUTTON.OLEMSGBUTTON_OK, OLEMSGDEFBUTTON.OLEMSGDEFBUTTON_FIRST);
+                    }
+                });
+
+            }
+        }
+
+        private bool HasMatchingSyntaxNode(SyntaxNode syntaxNode, string name)
+        {
+            if (syntaxNode is BaseTypeDeclarationSyntax || syntaxNode is DelegateDeclarationSyntax)
```

Comment:
This needs to be replicated for VB. Instead I suggest using SyntaxGenerator here so that it can remain language agnostic. Something like:

```csharp
var generator = SyntaxGenerator.GetGenerator(document);
var kind = generator.GetDeclarationKind(syntaxNode);

if (kind == DeclarationKind.Class || DeclarationKind.Interface || ...) 
{
     return generator.GetName(syntaxNode) == name;
}
```